### PR TITLE
Make layer() accept formula notation for data argument 

### LINKED ISSUE #3138: Support formula notation in `data` argument ?
The data argument can be used to pass a function to apply on the plot's data. It would be nice to be able to use the formula notation as well.

```
# works
ggplot(iris,aes(Sepal.Length,Sepal.Width)) + 
  geom_point(data = head)
# works as well
ggplot(iris,aes(Sepal.Length,Sepal.Width)) + 
  geom_point(data = function(x) head(x,10))
# doesn't
ggplot(iris,aes(Sepal.Length,Sepal.Width)) + 
  geom_point(data = ~head(.,10))
```

It should be a simple fix, either in `layer` with something like `if(inherits(data,"formula")) data <- rlang::as_function(data)` or in later in `ggproto` after defining `fortify.formula <- function(model, data, ...) model` to return the input as is done with `fortify.function`, with the same but also testing for missing with something like `if(!missing(data) && inherits(data,"formula")) data <- rlang::as_function(data)`

```
trace(ggplot2:::layer,at =2,quote(
  if(inherits(data,"formula")) data <- rlang::as_function(data)))   

# now works
ggplot(iris,aes(Sepal.Length,Sepal.Width)) + 
  geom_point(data = ~head(.,10))
```

### PR DESCRIPTION
**Pull request by yutannihilation**
Fix #3138

`data` is processed by `fortify()` once and stored in layer. Then, if it's a function, it's applied to the plot data when the plot is built,

https://github.com/tidyverse/ggplot2/blob/475042e9ce209e1279d77d46288315f801150d4d/R/layer.r#L191-L192

So, all we have to do is to define a `fortify()` method that converts a formula to a function.

## Usages

``` r
devtools::load_all("~/repo/R/ggplot2/")
#> Loading ggplot2

ggplot(iris, aes(Sepal.Length, Sepal.Width)) + 
  geom_point(data = ~ head(., 10))
```

<sup>Created on 2019-04-13 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

### CODE REVIEW

**Comment by thomasp85** [comment]
There's a dependency on #3242... If you merge that one first and then update this to remove the rlang namespacing it's all good :-)

**Comment by yutannihilation** [comment]
Sure, thanks!

File: man/theme.Rd [hunk 1]
```diff
@@ -251,15 +251,20 @@ p1 + theme(
 )
 
 # Axes ----------------------------------------------------------------------
+# Change styles of axes texts and lines
```
Comment (yutannihilation) [inline man/theme.Rd]: It seems these diffs of `theme.Rd` come from #2934. It might be better to create a separate PR, but I let this included.

Comment (clauswilke) [reply man/theme.Rd]: Actually, that PR also missed to generate a new vdiffr reference image. Do you have the time to create a separate PR that includes the updated .Rd file and the missing image? That would probably be the best approach.

Comment (yutannihilation) [reply man/theme.Rd]: Oh, sure. Will do.
